docs(rtd): include DFN specification in dev docs (#1948)
MODFLOW-USGS/modflow6: doc/mf6io/mf6ivar/readme.md

  - [Keystring](#keystring)
- [Text substitutions](#text-substitutions)
- [Writing definition files](#writing-definition-files)
  - [Subpackages](#subpackages)
  - [Models](#models)
  - [Solvers](#solvers)
- [Developer tooling](#developer-tooling)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## DFN file overview

A DFN file is a text file enumerating **input variables** for a MODFLOW 6 component. An input variable is any value provided by the user to configure a MODFLOW 6 simulation. Each definition file specifies the input file expected for some MODFLOW 6 component (e.g. a simulation, model, or package) and specifies zero or more variables.

Each variable is described by a set of attributes, some of which are required, some optional.

DFN files must contain only ASCII characters. DFN files may include comment lines, which must begin with "#".

Variables are organized into **blocks**. A block is a set of related input parameters. Each block in a DFN file is conventionally preceded by a comment line according to the format:

```
# --------------------- component subcomponent block ---------------------
```

> A variable is specified by a set of attributes on adjacent one on each line, one attribute per line, following the general format:

Reviewer: ```suggestion
A variable is specified by a set of attributes, with one attribute per line, following the general format:
```

```
name value
```

Some attribute values are optional. Variables are delimited by empty lines.

DFN files should be named `component-subcomponent.dfn`, where the component and subcomponent are abbreviated acronyms or names. For instance, a groundwater flow (GWF) model's initial conditions (IC) package DFN file is named `gwf-ic.dfn`.

DFN files are parsed to generate both source code and documentation.

**Note**: The DFN format is undergoing a migration to TOML. This involves transforming the flat variable representation into a tree with composite variables as nodes and scalars as leaves.

## Variable attributes

> A MODFLOW 6 input variable is described by a set of attributes. Some attributes are optional and do not need to be specified (because they have a default value) and others are required. The following attributes are supported:

Reviewer: Nice idea to write these as a table.

| deprecated    | Indicates that the parameter has been deprecated.                      | No       | None    | Should a semantic version number, the version in which the parameter was deprecated. If this attribute is provided without a value, it is ignored.            |

### Reader Attribute

The reader attribute indicates what reader is used by MODFLOW 6 for the information.  There are several reader types that result in specialized input instructions.  For example, the delr array of the DIS package is read using u1ddbl.  Because the MODFLOW 6 array readers often require a control record, when this reader type is specified, information about the control record is written.  For example, the following block identifies how delr is specified:

```
block disdata
name delr
type double precision
shape (ncol)
reader u1ddbl
longname spacing along a row
description is the is the column spacing in the row direction.
```

This results in the following block description:

```

> BEGIN DISDATA

Reviewer: This should be GRIDDATA, rather than DISDATA.

  DELR
    delr(ncol) -- U1DDBL
END DISDATA
```

The READARRAY reader is another reader that results in specialized input.  It allows for a LAYERED keyword to be specified.  The icelltype variable is read using readarray and is specified as:

```

> block npfdata

Reviewer: This block is GRIDDATA, right?

name icelltype
type integer
shape (nodes)
valid
reader readarray
layered true
optional
longname confined or convertible indicator
description flag for each cell that specifies how saturated thickness is treated.  0 means saturated thickness is held constant;  $>$0 means saturated thickness varies with computed head when head is below the cell top; $<$0 means saturated thickness varies with computed head unless the THICKSTRT option is in effect.  When THICKSTRT is in effect, a negative value of icelltype indicates that saturated thickness will be computed as STRT-BOT and held constant.
```

This results in the following block description:

```

> BEGIN NPFDATA

Reviewer: GRIDDATA
